### PR TITLE
write tests for merge_new_data

### LINKED ISSUE #6: finish the tests for merge_new_data
if you type "closes #6" in the top message field of the PR, then this will automatically close when that PR merges.

### PR DESCRIPTION
**Pull request by tinatn29**
closes #6 
Currently all tests in test_core.py except 'invalid_key' passed (when the existing dataset doesn't have the key found in new_dataset). Let me know if there are other cases I should add to test_core.py.

I need to edit the function to deal with 'invalid_key' cases. Should I do it on this branch and push again (so all the checks/tests pass?)

### CODE REVIEW

**Comment by tinatn29** [comment]
All checks passed! I modified merge_new_data to only merge if the key in new_dataset exists in self.dataset (I want to make this optional depending on a new function argument - will create a new issue + separate PR for that)

File: ml4ms/core.py [hunk 1]
```diff
@@ -9,21 +9,23 @@ def merge_new_data(self, new_data):
...
             nothing
         """
         for key, value in new_data.items():
-            self.dataset[key].update(value)
+            # merge only if key exists
+            if key in self.dataset:
+                self.dataset[key].update(value)
 
 
 if __name__ == "__main__":
```
Comment (sbillinge) [inline ml4ms/core.py:25]: I suggest we remove this.  Now we have tests we don't need it any more and we never expect anyone to actually run this module, only import things from it.

File: ml4ms/tests/test_core.py [hunk 2]
```diff
@@ -1,23 +1,42 @@
...
+    ),
     # test overlapping items in intersecting sets
-    #...
+    (
+        {"1": {"_id": 1, "a": 2}, "3": {"_id": 3}},
+        {"1": {"_id": 1, "a": 3, "b": 3}},  # this should overwrite a['1']
+        {"1": {"_id": 1, "a": 3, "b": 3}, "3": {"_id": 3}},
+    ),
+    # test merging an empty collection
+    ({"1": {"_id": 1, "a": 2}, "3": {"_id": 3}}, {}, {"1": {"_id": 1, "a": 2}, "3": {"_id": 3}}),
```
Comment (sbillinge) [inline ml4ms/tests/test_core.py:26]: maybe make this look like the others?

File: ml4ms/tests/test_core.py [hunk 3]
```diff
@@ -1,23 +1,42 @@
...
     # test the new dataset does not have the right primary key (invalid datasets could maybe be tested elsewhere)
-    # ...
+    (
+        {"1": {"_id": 1, "a": 2}, "3": {"_id": 3}},
+        {"invalid_key": {"_id": "invalid_key"}},
```
Comment (sbillinge) [inline ml4ms/tests/test_core.py:30]: I think this is a valid key. Wouldn't invalid be something like this:
`{"wrong_key: {"_id": "right_key"}}`

Though we may want to assume that this is checked when databases are loaded and it doesn't need to be checked here.

File: ml4ms/core.py [hunk 4]
```diff
@@ -9,21 +9,23 @@ def merge_new_data(self, new_data):
...
             {mat_id_1 : {attribute_1: , attribute_2:, ...}, ...}
 
         Returns
         -------
             nothing
         """
         for key, value in new_data.items():
-            self.dataset[key].update(value)
+            # merge only if key exists
+            if key in self.dataset:
```
Comment (sbillinge) [inline ml4ms/core.py:21]: As a rule I would advise against this as written as it introduces "magic" into the code.  Always try to minimize "magic".  Magic is when code behaves unexpectedly.  It is very frustrating when you encounter it.  It is much better to have the code crash if something unexpected is handed in than to have it skip over it and not tell anyone.

I saw in your comment you want to make this behavior  "optional".  Is there a reason for that?  Don't you want it always to just merge it?  It seems like a valid thing to want to add a new document that doesn't already exist.  What is the Use Case (UC) for handing a new document to the dataset but having it skipped over?

**Review by sbillinge — COMMENTED** [review]
this looks really great! Fantastic job.  

Please see the inline comments.  Also, did we lose the test for the orthogonal datasets, i.e., the case where the `new_data` doc is not in the `old_data`? update: I see that you are calling this test "invalid key" now.  I had something else in mind for invalid key....you will see what I meant in the inline comment.

I like your new structure, it is way easier to read, with old_data, new_data and expected.

**Pull request merged**